File: pascal/mandelbrot.pas
{ Compile for SDL1.2: fpc mandelbrot.pas }
{ Compile for SDL2: fpc -dSDL2 -FuPATH_TO_SDL2_UNIT mandelbrot.pas }
program mandelbrot;

uses ucomplex, math, sysutils, dateutils,
  {$ifdef unix}cthreads,{$endif}
  {$if defined(SDL2)} SDL2; {$else} sdl; {$endif}

const
   N                                = 1000;  { grid size }
   depth                            = 200;   { max iterations }
   escape2                          = 400.0; { escape radius ^ 2 }
   color_map: array[0..18, 0..2] of real = 
     (( 0.0, 0.0, 0.5 ) ,
      ( 0.0, 0.0, 1.0 ) ,
      ( 0.0, 0.5, 1.0 ) ,
      ( 0.0, 1.0, 1.0 ) ,
      ( 0.5, 1.0, 0.5 ) ,
      ( 1.0, 1.0, 0.0 ) ,
      ( 1.0, 0.5, 0.0 ) ,
      ( 1.0, 0.0, 0.0 ) ,
      ( 0.5, 0.0, 0.0 ) ,
      ( 0.5, 0.0, 0.0 ) ,
      ( 1.0, 0.0, 0.0 ) ,
      ( 1.0, 0.5, 0.0 ) ,
      ( 1.0, 1.0, 0.0 ) ,
      ( 0.5, 1.0, 0.5 ) ,
      ( 0.0, 1.0, 1.0 ) ,
      ( 0.0, 0.5, 1.0 ) ,
      ( 0.0, 0.0, 1.0 ) ,
      ( 0.0, 0.0, 0.5 ) ,
      ( 0.0, 0.0, 0.0 ) );

   parallelism = 16;
   jobsize = N*N div parallelism;

var
   min_result : real = 0.0; 
   max_result : real = 0.0;
   log_count  : array[0..(N*N-1)] of real;
   argb_array : array[0..(N*N-1)] of longword;

   threads : array [0..(parallelism-1)] of TThreadID;

{ ---------------------------------------------------------------------------- }

function trans_x(x: integer) : real; inline;
begin
   trans_x := 3.0 * (real(x) / N) - 2.0;
end;

{ ---------------------------------------------------------------------------- }

function trans_y(y: integer) : real; inline;
begin
   trans_y := 3.0 * (real(y) / N) - 1.5;
end;

{ ---------------------------------------------------------------------------- }

function mag2(const x: complex) : real; inline;
begin
   mag2 := x.re * x.re + x.im * x.im;
end;

{ ---------------------------------------------------------------------------- }

function mandel(idx : longword) : real; inline;
var
   z0    : complex;
   z     : complex;
   k     : integer;
   magz2 : real;
begin
   z0 := cinit(trans_x(idx mod N), trans_y(idx div N));
   z  := cinit(0.0, 0.0);
   k := 0;
   while k < depth do
   begin
      magz2 := mag2(z);
      if magz2 >= escape2 then
         break;
      z := z * z + z0;
      k := k + 1;
   end;
   mandel := log10(k + 1.0 - log10(log10(max(magz2, escape2)) / 2.0) / log10(2.0));
end;

{ ---------------------------------------------------------------------------- }

function interpolate(d : real; v0 : real; v1 : real) : integer; inline;
begin
   interpolate := trunc((d * (v1 - v0) + v0) * 255.0);
end;

{ ---------------------------------------------------------------------------- }

function map_to_argb(x :real) : longword; inline;
var
   bin                    : integer;
   r0, g0, b0, r1, g1, b1 : real;
   d                      : real;
   r,g,b                  : integer;
begin
   x := (x - min_result) / (max_result - min_result) * 18;
   bin := trunc(x);
   if (bin >= 18) then
      map_to_argb := $ff000000
   else
      begin
         r0 := color_map[bin][0];
         g0 := color_map[bin][1];
         b0 := color_map[bin][2];
         r1 := color_map[bin+1][0];
         g1 := color_map[bin+1][1];
         b1 := color_map[bin+1][2];
         d := x - bin;
         r := interpolate(d, r0, r1);
         g := interpolate(d, g0, g1);
         b := interpolate(d, b0, b1);
         map_to_argb := b or (g << 8) or (r << 16) or $ff000000;
      end;
end;

{ ---------------------------------------------------------------------------- }

function do_mandel(p : pointer) : ptrint;
var
   a, idx : longword;
begin
   for a := 0 to jobsize-1 do
   begin
      idx := a*parallelism+longword(p);
      log_count[idx] := mandel(idx);
   end;
   do_mandel := 0;
end;

{ ---------------------------------------------------------------------------- }

procedure do_map_to_argb;
var
   idx : longword;
begin
   max_result := maxvalue(log_count, N*N);
   min_result := minvalue(log_count, N*N);
   for idx := 0 to (N*N-1) do
      begin
         argb_array[idx] := map_to_argb(log_count[idx]);
      end;
end;

{ ---------------------------------------------------------------------------- }

var
{$if defined(SDL2)}
   win    : PSDL_Window;
   ren    : PSDL_Renderer;
   tex    : PSDL_Texture;
{$else}
   scr    : PSDL_Surface;
   row    : integer;
   col    : integer;
   line   : ^longword;
   stride : integer;
   idx    : longword;
{$endif}
   evt    : TSDL_Event;
   start  : TDateTime;
   finish : TDateTime;
   i      : integer;

begin
   start := now;
   for i := 0 to parallelism do
      threads[i] := beginthread(@do_mandel, pointer(i));
   for i := 0 to parallelism do
       waitforthreadterminate(threads[i], 0);
   do_map_to_argb();
   finish := now;
   writeln(MilliSecondSpan(start, finish):4:2, ' ms');

   SDL_init(SDL_INIT_VIDEO);

{$if defined(SDL2)}
   SDL_CreateWindowAndRenderer(N, N, 0, @win, @ren);
   SDL_SetRenderDrawColor(ren, 0, 0, 0, 255);
   SDL_RenderClear(ren);
   tex := SDL_CreateTexture(ren, SDL_PIXELFORMAT_ARGB8888, SDL_TEXTUREACCESS_STREAMING, N, N);
   SDL_UpdateTexture(tex, nil, @argb_array[0], N*4);
   SDL_RenderCopy(ren, tex, nil, nil);
   SDL_RenderPresent(ren);
{$else}
   scr := SDL_SetVideoMode(1000, 1000, 32, SDL_SWSURFACE);
   SDL_LockSurface(scr);
   line := scr^.pixels;
   stride := scr^.pitch div 4;   
   idx := 0;   
   for row := 1 to scr^.h do
      begin
         for col := 0 to scr^.w-1 do
            begin
               line[col] := argb_array[idx];
               idx := idx + 1;
            end;
         line := line + stride;
      end;   
   SDL_UnlockSurface(scr);
   SDL_Flip(scr);
{$endif}

   while SDL_WaitEvent(@evt) <> 0 do
   begin
      case evt.type_ of 
         SDL_KEYUP:     break;
         SDL_QUITEV:    break;
      end;              
   end;

   SDL_Quit;

end.

File: haskell/mandelbrot.hs

import Control.Monad
import System.IO
import System.CPUTime
import Text.Printf
import Data.Complex
import Data.Bits
import Data.Binary
import Text.Printf

import qualified Data.ByteString as BS

parallelism = 16

-- ********************************************************************
-- Mandelbrot

grid_dim = 100::Int        -- grid size (N)

depth = 200::Int           -- max iterations

escape2 = 400.0            -- escape radius ^ 2

trans_x :: Int -> Double
trans_x x =
  3.0*fromIntegral x/fromIntegral grid_dim - 2.0

trans_y :: Int -> Double
trans_y y =
  3.0*fromIntegral y/fromIntegral grid_dim - 1.5

fm :: Int -> Complex Double -> Complex Double
fm k z0
  | k <= 0    = z0
  | otherwise = (fm (k-1) z0)^2 + z0

mandel :: Int -> Double
mandel idx =
  let z0 = (trans_x (idx `mod` grid_dim)) :+ (trans_y (idx `div` grid_dim))
      k = [0..depth]
      zs = map (\k -> fm k z0) k
      zmags = map magnitude zs
      zmags' = takeWhile (< escape2) zmags
      magz2 = maximum zmags'
      kn = length zmags'
      v = max magz2 escape2
  in
    log (fromIntegral kn + 1.0 - log (log v / 2.0) / log 2.0)

-- ********************************************************************
-- Color mapping
color_map :: [[Double]]
color_map =
    [ [ 0.0, 0.0, 0.5 ] ,
      [ 0.0, 0.0, 1.0 ] ,
      [ 0.0, 0.5, 1.0 ] ,
      [ 0.0, 1.0, 1.0 ] ,
      [ 0.5, 1.0, 0.5 ] ,
      [ 1.0, 1.0, 0.0 ] ,
      [ 1.0, 0.5, 0.0 ] ,
      [ 1.0, 0.0, 0.0 ] ,
      [ 0.5, 0.0, 0.0 ] ,
      [ 0.5, 0.0, 0.0 ] ,
      [ 1.0, 0.0, 0.0 ] ,
      [ 1.0, 0.5, 0.0 ] ,
      [ 1.0, 1.0, 0.0 ] ,
      [ 0.5, 1.0, 0.5 ] ,
      [ 0.0, 1.0, 1.0 ] ,
      [ 0.0, 0.5, 1.0 ] ,
      [ 0.0, 0.0, 1.0 ] ,
      [ 0.0, 0.0, 0.5 ] ,
      [ 0.0, 0.0, 0.0 ] ]

interpolate :: Double -> Double -> Double -> Int
interpolate d v0 v1 =
  round ((d * (v1 - v0) + v0) * 255.0)

stops = 18

map_to_argb :: Double -> Double -> Double -> Int
map_to_argb min_result max_result x =
  let
    x1 = (x - min_result) / (max_result - min_result) * (fromIntegral stops)
    bin = round x1
  in
    if bin >= stops
    then 0xff000000
    else let
          r0 = color_map !! (fromIntegral bin) !! 0
          g0 = color_map !! (fromIntegral bin) !! 1
          b0 = color_map !! (fromIntegral bin) !! 2
          r1 = color_map !! (fromIntegral (bin+1)) !! 0
          g1 = color_map !! (fromIntegral (bin+1)) !! 1
          b1 = color_map !! (fromIntegral (bin+1)) !! 2
          d = x1 - (fromIntegral bin)
          r = interpolate d r0 r1
          g = interpolate d g0 g1
          b = interpolate d b0 b1
        in
          b + (g * 256) + (r * 65536) + 0xff000000

job_size = grid_dim*grid_dim `div` parallelism

worker_threads = parallelism

do_mandel =
  let
    r = [0 .. (grid_dim*grid_dim)]
  in
   map mandel r

do_map_to_argb :: [Double] -> [Int]
do_map_to_argb log_count =
  let
    min_result = minimum log_count
    max_result = maximum log_count
  in
   map (map_to_argb min_result max_result) log_count

-- Converts ARGB32 into RGB24
hPutRGB :: Handle -> Int -> IO ()
hPutRGB hfile word =
  let
    r = fromIntegral ((word .&. 0x00ff0000::Int) `shiftR` 16)::Word8
    g = fromIntegral ((word .&. 0x0000ff00::Int) `shiftR`  8)::Word8
    b = fromIntegral ((word .&. 0x000000ff::Int))::Word8
    bs = BS.pack [r,g,b]
  in
    BS.hPut hfile bs

saveAsPPM :: [Int] -> String -> IO ()
saveAsPPM bin filename = do
  outf <- openFile filename WriteMode
  hPutStrLn outf "P6"
  hPutStrLn outf $ printf "%u %u" grid_dim grid_dim
  hPutStrLn outf "255"
  mapM_ (hPutRGB outf) bin
  hClose outf

time :: IO t -> IO t
time a = do
    start <- getCPUTime
    v <- a
    end   <- getCPUTime
    let diff = (fromIntegral (end - start)) / (10^12)
    printf "Computation time: %0.3f sec\n" (diff :: Double)
    return v

process =
  let
    log_count = do_mandel
    argb_array = do_map_to_argb log_count
  in do
    saveAsPPM argb_array "output.ppm"

main = do
  putStrLn "mandelbrot"
  time process
